precompiles/assets/src/lib.rs

```diff
@@ -0,0 +1,440 @@
+			SELECTOR_LOG_TRANSFER,
+			H160::default(),
+			to,
+			EvmDataWriter::new().write(value).build(),
+		)
+		.record(handle)?;
+
+		Ok(true)
+	}
+
+	#[precompile::public("burn(address,uint256)")]
+	fn burn(handle: &mut impl PrecompileHandle, from: Address, value: U256) -> EvmResult<bool> {
+		handle.record_log_costs_manual(3, 32)?;
+		let asset_id = Self::asset_id(handle)?;
+
+		let from: H160 = from.into();
+		let value = Self::u256_to_amount(value).in_field("value")?;
+		{
+			let origin: AccountIdOf<Runtime> = handle.context().caller.into();
+			let from: AccountIdOf<Runtime> = from.clone().into();
+
+			RuntimeHelper::<Runtime>::try_dispatch(
+				handle,
+				Some(origin).into(),
+				pallet_assets::Call::<Runtime>::burn {
+					id: asset_id,
+					who: Runtime::Lookup::unlookup(from),
+					amount: value,
+				},
+			)?;
+		}
+
+		log3(
+			handle.context().address,
+			SELECTOR_LOG_TRANSFER,
+			from,
+			H160::default(),
+			EvmDataWriter::new().write(value).build(),
+		)
+		.record(handle)?;
+
+		Ok(true)
+	}
+
+	#[precompile::public("transfer_ownership(address)")]
+	fn transfer_ownership(handle: &mut impl PrecompileHandle, owner: Address) -> EvmResult<bool> {
+		let asset_id = Self::asset_id(handle)?;
+
+		let owner: H160 = owner.into();
+		{
+			let origin: AccountIdOf<Runtime> = handle.context().caller.into();
+			let owner: AccountIdOf<Runtime> = owner.into();
+
+			RuntimeHelper::<Runtime>::try_dispatch(
+				handle,
+				Some(origin).into(),
+				pallet_assets::Call::<Runtime>::transfer_ownership {
+					id: asset_id,
+					owner: Runtime::Lookup::unlookup(owner),
+				},
+			)?;
+		}
+
+		Ok(true)
+	}
+
+	#[precompile::public("freeze(address)")]
+	fn freeze(handle: &mut impl PrecompileHandle, account: Address) -> EvmResult<bool> {
+		let asset_id = Self::asset_id(handle)?;
+
+		let account: H160 = account.into();
+		{
+			let origin: AccountIdOf<Runtime> = handle.context().caller.into();
+			let account: AccountIdOf<Runtime> = account.into();
+
+			RuntimeHelper::<Runtime>::try_dispatch(
+				handle,
+				Some(origin).into(),
+				pallet_assets::Call::<Runtime>::freeze {
+					id: asset_id,
+					who: Runtime::Lookup::unlookup(account),
+				},
+			)?;
+		}
+
+		Ok(true)
+	}
+
+	#[precompile::public("thaw(address)")]
+	fn thaw(handle: &mut impl PrecompileHandle, account: Address) -> EvmResult<bool> {
+		let asset_id = Self::asset_id(handle)?;
+
+		let account: H160 = account.into();
+		{
+			let origin: AccountIdOf<Runtime> = handle.context().caller.into();
+			let account: AccountIdOf<Runtime> = account.into();
+
+			RuntimeHelper::<Runtime>::try_dispatch(
+				handle,
+				Some(origin).into(),
+				pallet_assets::Call::<Runtime>::thaw {
+					id: asset_id,
+					who: Runtime::Lookup::unlookup(account),
+				},
+			)?;
+		}
+
+		Ok(true)
+	}
+
+	fn asset_id(handle: &mut impl PrecompileHandle) -> EvmResult<AssetIdOf<Runtime>> {
+		let asset_id = AssetIdConverter::account_to_asset_id(handle.code_address().into());
```

Comment:
We should check the prefix of the account id? The prefix should be zero,

Otherwise, non zero prefix could be truncated, and not detected.

Reply:
I am not sure I entirely understand the point you said here.  Assume that the `code_address` is not a zero prefix address like `0x6Be02d1d3665660d22FF9624b7BE0551ee1Ac91b`, then the `asset_id` generated here would be a much large number according to the `AccountToAssetId` rule defined in runtime side.  Then we'll try to fetch the asset info in the asset pallet, If it does not exist, return the error directly.

```rs
if pallet_assets::Pallet::<Runtime>::maybe_total_supply(asset_id).is_some() {
	return Ok(asset_id);
}
Err(revert("The asset not exist!"))
```

On the runtime side, Only `AssetIdConverter::account_to_asset_id(a.into())` located in the `(1026..1536)` range will continue the precompile execution, otherwise, return None.

```rs
fn execute(&self, handle: &mut impl PrecompileHandle) -> Option<PrecompileResult> {
	match handle.code_address() {
                a if (1026..1536).contains(&AssetIdConverter::account_to_asset_id(a.into())) 
                      =>Some(<ERC20Assets<Runtime, AssetIdConverter>>::execute(handle)),
        ｝
}
```